File: app/src/main/java/closer/vlllage/com/closer/handler/helpers/MenuOptionAdapter.java
package closer.vlllage.com.closer.handler.helpers;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import closer.vlllage.com.closer.R;

class MenuOptionAdapter extends RecyclerView.Adapter<MenuOptionAdapter.ViewHolder> {

    private List<MenuHandler.MenuOption> menuOptions;
    private OnMenuOptionClickListener onMenuOptionClickListener;

    public MenuOptionAdapter(
            @NonNull List<MenuHandler.MenuOption> menuOptions,
            @NonNull OnMenuOptionClickListener onMenuOptionClickListener) {
        this.menuOptions = menuOptions;
        this.onMenuOptionClickListener = onMenuOptionClickListener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int position) {
        return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.menu_modal_item, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder viewHolder, int position) {
        MenuHandler.MenuOption menuOption = menuOptions.get(position);
        viewHolder.name.setText(menuOption.titleRes);
        viewHolder.name.setCompoundDrawablesWithIntrinsicBounds(menuOption.iconRes, 0, 0, 0);
        viewHolder.itemView.setOnClickListener(view -> onMenuOptionClickListener.onMenuOptionClick(menuOption));
    }

    @Override
    public int getItemCount() {
        return menuOptions.size();
    }

    static class ViewHolder extends RecyclerView.ViewHolder {

        TextView name;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            name = itemView.findViewById(R.id.name);
        }
    }

    interface OnMenuOptionClickListener {
        void onMenuOptionClick(MenuHandler.MenuOption menuOption);
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/helpers/MenuOptionAdapter.kt
package closer.vlllage.com.closer.handler.helpers

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import closer.vlllage.com.closer.R

internal class MenuOptionAdapter(
        private val menuOptions: List<MenuHandler.MenuOption>,
        private val onMenuOptionClickListener: (menuOption: MenuHandler.MenuOption) -> Unit) : RecyclerView.Adapter<MenuOptionAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, position: Int): ViewHolder {
        return ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.menu_modal_item, parent, false))
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
        val menuOption = menuOptions[position]
        menuOption.titleRes?.let { viewHolder.name.setText(it) }
        menuOption.title?.let { viewHolder.name.text = it }
        viewHolder.name.setCompoundDrawablesWithIntrinsicBounds(menuOption.iconRes, 0, 0, 0)
        viewHolder.itemView.setOnClickListener { onMenuOptionClickListener.invoke(menuOption) }
    }

    override fun getItemCount() = menuOptions.size

    internal class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var name: TextView = itemView.findViewById(R.id.name)
    }

}
